modules/table/src/main/java/org/apache/ignite/internal/table/distributed/storage/InternalTableImpl.java

```diff
@@ -2098,14 +2106,24 @@ protected CompletableFuture<ClusterNode> evaluateReadOnlyRecipientNode(int partI
                         throw withCause(TransactionException::new, REPLICA_UNAVAILABLE_ERR, e);
                     } else {
                         if (res == null) {
-                            throw withCause(TransactionException::new, REPLICA_UNAVAILABLE_ERR, e);
+                            throw createTransactionException(tablePartitionId, readTimestamp);
                         } else {
                             return getClusterNode(res);
                         }
                     }
                 });
     }
 
+    private static TransactionException createTransactionException(TablePartitionId tablePartitionId, HybridTimestamp readTimestamp) {
```

Comment:
A very general name for a method, it can be misused, it is better to give it a name like "createFailedGetPromaryReplicaTransactionException".